dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java

```diff
@@ -17,156 +17,342 @@
+
+    checkArgument(optionalReferencedFileType.isPresent(), "Referenced file type must be provided for '%s' error "
+        + "type", KVErrorType.class.getSimpleName());
+
+    if (errorType == RELATION) {
+      val errorFileds = relationsStream
+          .filter(relation -> !hasOptionalFk(relation, fileSchema))
+          .map(relation -> getFileTypeFKs(relation, fileSchema))
+          .findFirst();
+      checkState(errorFileds.isPresent(), "Failed to resolve error fields for file type '%s'; error type '%s' and "
+          + "referenced file type '%s'", fileType, errorType, referencedFileType);
+
+      return errorFileds.get();
+    }
+
+    if (errorType == OPTIONAL_RELATION) {
+      val errorFileds = relationsStream
+          .filter(relation -> hasOptionalFk(relation, fileSchema))
+          .map(relation -> getOptionalFileTypeFKs(relation, fileSchema))
+          .findFirst();
+      checkState(errorFileds.isPresent(), "Failed to resolve error fields for file type '%s'; error type '%s' and "
+          + "referenced file type '%s'", fileType, errorType, referencedFileType);
+
+      return errorFileds.get();
+    }
+
+    if (errorType == CONDITIONAL_RELATION) {
+      val errorFileds = relationsStream
+          .filter(relation -> hasConditionalFk(relation))
+          .map(relation -> getConditionalFileTypeFKs(relation, fileSchema))
+          .findFirst();
+      checkState(errorFileds.isPresent(), "Failed to resolve error fields for file type '%s'; error type '%s' and "
+          + "referenced file type '%s'", fileType, errorType, referencedFileType);
+
+      return errorFileds.get();
+    }
+
+    throw new IllegalArgumentException(format("Failed to resolve error filed names for error type '%s'", errorType));
+  }
+
   @Override
-  public Optional<KVFileType> getOptionalReferencedFileType2(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return null;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getErrorFieldNames(org.icgc.dcc.submission.validation.
-   * key.core.KVFileType, org.icgc.dcc.submission.validation.key.core.KVErrorType)
-   */
+  public List<String> getPrimaryKeyNames(@NonNull KVFileType fileType) {
+    val fileSchema = getFileSchema(fileType);
+
+    return fileSchema.getUniqueFields().stream()
+        .sorted()
+        .collect(toImmutableList());
+  }
+
   @Override
-  public List<String> getErrorFieldNames(KVFileType fileType, KVErrorType errorType) {
-    // TODO Auto-generated method stub
-    return null;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getPrimaryKeyNames(org.icgc.dcc.submission.validation.
-   * key.core.KVFileType)
-   */
+  public Iterable<KVFileType> getTopologicallyOrderedFileTypes() {
+    return KVDynamicDictionaryHelper.getTopologicallyOrderedFileTypes(dictionary);
```

Comment:
Not a fan of `Helper` classes as it is a `Utils` in general and thus typically has poor cohesion.